test/tests/pbs_qstat.py

```diff
@@ -0,0 +1,74 @@
+# coding: utf-8
+
+# Copyright (C) 1994-2018 Altair Engineering, Inc.
+# For more information, contact Altair at www.altair.com.
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# PBS Pro is free software. You can redistribute it and/or modify it under the
+# terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from ptl.utils.pbs_testsuite import *
+
+
+class QstatTest(PBSTestSuite):
+
+    """
+    This test suite checks whether qstat sends the
+    status request to the correct server or not for
+    requests like job_id.serverA@serverB
+    """
+
+    def setUp(self):
+        if len(self.servers.keys()) != 2:
+            self.skipTest("This needs two servers")
+        self.server1 = self.servers.values()[0]
+        self.server2 = self.servers.values()[1]
+        a = {'scheduling': 'false', 'flatuid': 'true'}
+        self.server1.manager(MGR_CMD_SET, SERVER, a)
+        self.server2.manager(MGR_CMD_SET, SERVER, a)
+        self.du.set_pbs_config(self.server.client, confs={'PBS_SERVER': ''})
+
+    def test_qstat_req_server(self):
+        j = Job(TEST_USER)
+        actualclient = self.server.client
+        self.server.client = self.server1.hostname
+        jid = self.server1.submit(j)
+        self.server1.expect(JOB, {'job_state': 'Q'}, id=jid)
+        destination = '@%s' % self.server2.hostname
+        self.server1.movejob(jobid=jid, destination=destination)
+        self.server.client = actualclient
+        # Allowing 5 secs of time difference in server and client
+        starttime = time.time()-5
```

Comment:
could you please explain the reason behind this time difference?

Reply:
When I was writing and running the test  every time I was experiencing 1 or 2 seconds difference in the client and server while doing qstat, so I thought allowing the 5 seconds time difference would be a good Idea.

Reply:
If I understand correctly you are saying that there is a time difference of 1-2 seconds between two different machines, which may or may not always happen. It is also possible that 5 seconds may not be enough if the time difference is more than 5 seconds.

Reply:
Yes @prakashcv13, I was seeing a delay if I was not using 5 second of difference because while I was doing qstat it was checking for the logs from the start time and later but in the server it was printing just a second before so I considered 5 seconds time difference. If the difference is more than that the test will fail and one needs to debug in such case ss there any time limit that we can consider for the difference between two machines?

Or we can also check for the logs by using tail, by this we can check the server log for the expected message from the bottom to top by mentioning some range. What you suggest?

Reply:
ok. In that case, how will you make sure that the "Type 19" being matched is not for the status request on line 64?

Reply:
@prakashcv13 Before movejob Type 19 request will go only in server1's log file because request if going to server1 not server2 so there if not Type 19 request log in server2's log file and once job gets moved then Type 19 request will go in server2's log file not in server1's log file as request if going to server2 not server1.
So at line 64 (which expect) will go to server1 and this status will go to server2.
Hope this what you are looking for.

Reply:
Yes, but I am also worried that type 19 log does not include the job id. How can we be sure that the type 19 if for the job we are looking for. Are we assuming that the no other client will be sending any stat request to server2 when the test is running?

Reply:
Yes, as per the test case only one client is there and that will make the request.

Reply:
well, it is an assumption that only this test case will be running in the complex. The setup I am referring to is where there are other clients apart from the client used by this test suite.

Reply:
In the log match with type 19 request we are also checking which client has requested.

Reply:
There is a possibility that a status request is sent from outside the test case on the same client for another job id. The test case will still pass.
I do agree that such a thing will not happen in the ideal world.